src/sinks/util/request_builder.rs

```diff
@@ -67,6 +67,10 @@ pub trait RequestBuilder<Input> {
         &self,
         events: Self::Events,
     ) -> Result<EncodeResult<Self::Payload>, Self::Error> {
+        // TODO: Should we add enough bounds on `Self::Events` that we could automatically derive event count/event byte
+        // size, and then we could generate `BatchRequestMetadata` and pass it directly to `build_request`? That would
+        // obviate needing to wrap `payload` in `EncodeResult`, although practically speaking.. the name would be kind
+        // of clash-y with `Self::Metadata`.
```

Comment:
I like the idea of automatically generating the data. I think we could handle the name clash if the results are worth it. In any case, noting this TODO needs a follow up.

Reply:
Yeah, to say it loud so it's not just a thought in my brain: I think `RequestMetadata` could actually take on a generic parameter for a `data` field, _and_ a field for finalizers.

Right now, `RequestBuilder` bifurcates the inputs into metadata and events, and then encodes the events, and passes the encoded events and the bifurcated metadata to the method that inevitably builds the request.

Instead, we could make it so that the metadata/event splitting step stays as-is, but the middle step generates the count/size stuff, wraps the user-provided metadata, and then passes the wrapped metadata (`RequestMetadata<Self::Metadata>`) and the payload at the very end.

As far as the finalizers go, we basically always factor them out of the input events and pass them within `Self::Metadata`, so we might as well just do that for the implementor as well. This would also let us implement `Finalizable` for `RequestMetadata<T>`, removing yet another needlessly-broken-out field on the "true" request value i.e. `GcsRequest`.

Reply:
Opened #12857 to track this.